lib/phoenix/router/helpers.ex

```diff
@@ -8,6 +8,41 @@ defmodule Phoenix.Router.Helpers do
 
   @transports [Phoenix.Transports.WebSocket, Phoenix.Transports.LongPoller]
 
+  @doc false
+  def build_path(router, %Conn{} = conn, path) do
+    conn
+    |> build_own_forward_path(router, path)
+    |> Kernel.||(build_conn_forward_path(conn, router, path))
+    |> Kernel.||(path_with_script(path, conn.script_name))
+  end
+
+  defp build_own_forward_path(conn, router, path) do
+    case Map.fetch(conn.private, router) do
+      {:ok, {local_script, _}} ->
+        path_with_script(path, local_script)
+      :error -> nil
+    end
+  end
+
+  defp build_conn_forward_path(conn, router, path) do
+    case Map.fetch(conn.private, conn.private[:phoenix_router]) do
```

Comment:
In case `conn.private[:phoenix_router]` returns nil and there is a cruel soul that stores nil in `conn.private`, I would explicitly check if phoenix_router is actually set. You can even pattern match on it in the function head.